[linux] add build time option to disable automatic addon configure at…

wsnipex opened the pull request:
… startup

## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
Adds a cmake compile time option  `ADDONS_NO_CONFIGURE_AT_STARTUP`  to disable automatic enable/configure of binary addons on startup.
Fixes crashes with (bundled) pvr addons on e.g. flatpak.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Allows us to get rid of patches such as https://github.com/flathub/tv.kodi.Kodi/blob/master/addons-no-startupenable.patch
IIRC, LE uses something similar.


## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

wsnipex committed:
[linux] add build time option to disable automatic addon configure at startup

howie-f reviewed (approved):
this makes perfect sense. get rid of patches.
how would LE / flatpak activate this switch? just `cmake -DADDONS_NO_CONFIGURE_AT_STARTUP` ? no

wsnipex commented:
cmake -DADDONS_NO_CONFIGURE_AT_STARTUP=ON

garbear commented:
I like this feature, as I've had to hack this line when doing embedded Kodi stuff like my smarthome project, where I don't want anything to run other than my camera window or train HUD.

However, I would remove the need for the double-negative (and have the option default to YES):

```
cmake -DADDONS_CONFIGURE_AT_STARTUP=YES
```

wsnipex commented:
updated as per @garbear suggestion.
Renamed the option to ADDONS_CONFIGURE_AT_STARTUP, which defaults to ON.
Use -DADDONS_CONFIGURE_AT_STARTUP=OFF to disable

howie-f commented on CMakeLists.txt:
```diff
@@ -83,6 +83,7 @@ if(NOT WIN32)
 endif()
 if(CORE_SYSTEM_NAME STREQUAL linux)
   option(ENABLE_EVENTCLIENTS    "Enable event clients support?" OFF)
+  option(ADDONS_CONFIGURE_AT_STARTUP "Configure binary addons at startup?" ON)
```
alright, i missed the default in the first commit, thus the question how le and fp would activate :)

garbear commented on xbmc/platform/linux/PlatformLinux.h:
```diff
@@ -23,7 +23,9 @@ class CPlatformLinux : public CPlatformPosix
 
+#if defined(ADDONS_CONFIGURE_AT_STARTUP)
   bool IsConfigureAddonsAtStartupEnabled() override { return true; }
+#endif
```
I don't like how the if-defery affects the API inheritance. I would prefer to not override conditionally, and encapsulate the if-defery in the function:

Header:

```c++
bool IsConfigureAddonsAtStartupEnabled() override;
```

Cpp:

```c++
bool CPlatformLinux::IsConfigureAddonsAtStartupEnabled()
{
#if defined(ADDONS_CONFIGURE_AT_STARTUP)
  return true;
#else
  return false;
#endif
}
```

wsnipex commented:
shed repainted once more, good now @garbear?

eigendude reviewed (approved):
I try to bike shed when I don't have anything critical to say. Basically the PR was excellent so I had to nitpick something. And it's been nice seeing style recommendations being embraced and popping out some high quality code.

garbear reviewed (approved):
Sorry phone was logged into my work account. Looks great!

howie-f commented:
iirc it‘s not really crashing but you get the `would you like to enable?` question for any bundled addon. choose `yes` and the configuration dialog opens. this behavior is not appropiate for flatpak and le (where we have bundled dvb addons also)

ksooo commented:
@howie-f may I ask what is this configure at startup feature is good for?

howie-f commented:
after `apt install kodi-pvr-hts` on debian/ubuntu… or after an addon was installed manually to the filesystem on os level - kodi will ask you to enable and configure on next start. it was a wanted feature for distributions. but caused problems with libreelec dvb addons.

it was simply patched until now
https://github.com/flathub/tv.kodi.Kodi/blob/master/addons-no-startupenable.patch

howie-f commented:
means now drop the patch and build with ` -DADDONS_CONFIGURE_AT_STARTUP=YES ` is the same

ksooo commented:
So the actual problem is that we cannot distinguish between user installed (apt install, manual copy to addon system directory) vs. bundled sind. THIS is the root cause and that would need to be fixed. This PR here just works around the actual problem.

howie-f commented:
that is correct, but the only idea i have for now is an additional xml, let‘s call it `bundled-addons.xml` or whatever. ideally all platforms included, because it is different for UWP, tvOS (that’s why the feature is disabled on these platforms). flatpak and libreelec would need to maintain their own xml. hm, i am not really convinced. i‘m all ears for a better approach of course

ksooo commented:
We need a dedicated directory for bundled addons, which is different from system (aka all users) and user. At least this is how other products for this problem.
